boolean widget & some documentation

daddykotex opened the pull request:
**- Summary**

While working on the file-system backend, I was working with `hugo`. `hugo` uses a draft attribute to let you mark content as draft so it's not included in the output. I found there was no `boolean` widget, so here it is.

The MR passes `npm run lint:staged` but I have issues the the pre-commit hooks:
```
.git/hooks/pre-commit: line 2: ./node_modules/pre-commit/hook: No such file or directory
```

**- Test plan**

There's no test for the other widget, so I did not really know where to start here. Let me know if you want one.

<img width="833" alt="screen shot 2017-05-01 at 21 41 41" src="https://cloud.githubusercontent.com/assets/5230460/25601150/df19821a-2eb7-11e7-97fb-590f5f4d070c.png">

**- Description for the changelog**
Widget for `boolean` fields

erquhart commented:
@daddykotex great addition! Also love the bonus docs improvement, thanks for that.

Two thoughts:

1. Are you up for improving this to use the React Toolbox [switch widget](http://react-toolbox.com/#/components/switch)? The dependency is already there, and it's a very simple setup.
2. I realize other widgets such as the Number widget are using classes for no reason, but we want to use functional components wherever possible moving forward.

daddykotex commented:
Something like:

<img width="649" alt="screen shot 2017-05-02 at 22 20 27" src="https://cloud.githubusercontent.com/assets/5230460/25646359/9b0bc672-2f85-11e7-98c7-a36eb07c9cd8.png">

One thing I'm unsure is the behavior  if `value` is undefined, like if `required: false` and `draft` is not in the front matter for example. Right now, I think there's a stack in the console.

Benaiah commented:
@daddykotex I think the best way to handle this would be to require a default in the widget declaration in `config.yml`, and then have missing or undefined values default to that.

biilmann commented:
<img width="904" alt="screen shot 2017-05-03 at 2 14 20 pm" src="https://cloud.githubusercontent.com/assets/6515/25681906/eaec36ba-300a-11e7-9513-c16123cc040d.png">

There's still some UX issues with some of the widgets

Benaiah commented:
@biilmann did you mean to make that comment on @rafaelconde's PR? I'm not following how it's related to this one.

daddykotex commented:
@Benaiah that's a good idea, I'll play around to see how to implement that ;)

daddykotex commented:
Ok, I turned it functionnal (AFAIK 😄 ), it uses the switch widget and falls back to the default value or false if not supplied.

But now, I can't save because the validation proccess in `ControlHOC.js` seems to use `refs` and react gives `undefined` as `wrappedControl` in the function `processInnerControlRef`.

It blows up on save because validation is triggered and `const response = this.wrappedControlValid();` is called and at this point, it's undefined for the boolean controller.

Do you have any idea?

daddykotex commented:
Any input here?

daddykotex commented:
This implementation works correctly: https://github.com/daddykotex/netlify-cms/commit/eb1538c98ee561f4aca2f0e4a60e7802cb77b06d

Benaiah commented:
@daddykotex that commit does appear to be working the first time you edit a post, but afterwards I can't open any entry for editing. I'm looking into the issue.

daddykotex commented:
OK, so I'll try to be really clear so we avoid wasting time due to confusion. 

When I use the class implementation (bool-class branch https://github.com/daddykotex/netlify-cms/tree/bool-class). It works fine. I can save the `boolean` value, I can have a default value as well.

When I use the fully functional approach (bool branch https://github.com/daddykotex/netlify-cms/tree/bool), it does not work properly. I can't save and this is probably due to what I explained above (related to ref and `ControlHOC`). I see the following stack on save:

<img width="867" alt="screen shot 2017-05-13 at 09 43 53" src="https://cloud.githubusercontent.com/assets/5230460/26025986/d23ce098-37c0-11e7-8596-27d05e1f4455.png">

By the way, I rebased both branch on `master` and I updated the `example/config.yml` to have a `draft` using the boolean widget. This way, it might be wasier to test for you.

------------


On another note, I see that on FF and Chrome, the included example has a slightly different behavior. On Firefox, the markdown widget has a switch turned on while it's off on chrome:

<img width="1668" alt="screen shot 2017-05-13 at 09 36 10" src="https://cloud.githubusercontent.com/assets/5230460/26025996/2a381dc6-37c1-11e7-9efd-e7b9d7ca7a10.png">

I think this is completely unrelated though.

erquhart commented:
@daddykotex functional components don't have refs, so class is fine.

daddykotex commented:
Ok, so the implementation now uses a `class`.

Benaiah commented on example/config.yml:
```diff
@@ -15,6 +15,7 @@ collections: # A list of collections the CMS should be able to edit
       - {label: "Publish Date", name: "date", widget: "datetime", format: "YYYY-MM-DD hh:mma"}
       - {label: "Cover Image", name: "image", widget: "image", required: false, tagname: ""}
       - {label: "Body", name: "body", widget: "markdown"}
+      - {label: "Draft", name: "draft", widget: "boolean", required: false, defaultValue: true}
```
We don't want to use the example config as a test case for new features - can this line be removed?

daddykotex replied:
will do!

daddykotex commented:
removed the line and rebase against master :)

erquhart commented:
Simplified just a bit.
